Update The One Reference

jelaniwoods opened the pull request:
firstdraft/appdev#154

All methods should have:

- method name
- method signature
- short description
- return value class (link to relevant section of The One Reference)
- examples
- related methods
- link to full explanation in chapters if available

Currently, the extra loop methods do not have any descriptions in chapters so they do not have links.
`upto`, `downto`, `step`

The description for the `belongs_to` and `has_many` have not been modified since they don't really exist in another place.

Still needs some work to be done adding related methods links and return value Class links.

pmckernin commented on the-one-reference.md:
```diff
@@ -441,426 +445,834 @@ When given a string literal as an argument, it adds that string to the original
+
+Returns
+
+`false`
```
Change to 'true'

pmckernin commented on the-one-reference.md:
````diff
@@ -441,426 +445,834 @@ When given a string literal as an argument, it adds that string to the original
+
+```ruby
 lucky_number = rand(10) # assigns a random integer between 0 to 9 to the variable lucky_number
 p "My lucky number is " + lucky_number.to_i + "!" 
````
lucky_number.to_s

pmckernin commented on the-one-reference.md:
```diff
@@ -441,426 +445,834 @@ When given a string literal as an argument, it adds that string to the original
+
+### Subtraction
+
+Two dates can be subtracted from one another. The `-` operator returns a `Rational` which can be converted into an [`Integer`](#integer) to find the days in between the two dates.  
```
Should we define what a 'rational' is?

pmckernin commented on the-one-reference.md:
```diff
@@ -441,426 +445,834 @@ When given a string literal as an argument, it adds that string to the original
+
+`Array#push(Object)`
+
+Adds elements to an Array. Returns the modified Array.
```
I don't know if it matters, but should we say adds to the end of an array.  There are ways to prepend

pmckernin commented on the-one-reference.md:
````diff
@@ -441,426 +445,834 @@ When given a string literal as an argument, it adds that string to the original
-```
+### .store
+
+`Hash#store(Object, Object)`
````
Does it make more sense to have 'Hash#store(_key_, _value_)?

jelaniwoods replied:
@pmckernin The argument for using `Object` was because the signature should tell you the type of Object that the method accepts. We talked a little more about that here: firstdraft/appdev#154

raghubetina commented:
@jelaniwoods I know this has been merged already, but a few notes:

 - What do you think about putting the class of the return value more front-and-center, like part of the signature?
 - The `ActiveRecord_Relation` methods like `.where`, etc, are technically _instance_ methods I think. They are shown with a `.`, as opposed to something like `String#downcase`?

raghubetina commented:
"When given a string literal as an argument, it adds that string to the original string. .+ or + is shorthand for .concat method. Each line of code below will give the same output."

I don't think this is technically true; I believe `concat` alters the object in place. We could maybe change this to ".+ is _similar_ to concat", and/or break out `+` into its own section?

jelaniwoods commented:
@raghubetina I think that could be nice. Do you mean something like:
```ruby
String#split(String) ⇒ Array
```

Yes, that _is_ true about `ActiveRecord_Relation` methods. I guess I was thinking about how we used to say to the students that `where` is a method you call on a Class and not a record, but that's not entirely true. I'll update that.

Oh! That's interesting. I assumed `concat` and `+` were the same but they definitely are not. I think breaking them out into two separate listings is a good idea.

raghubetina commented:
> String#split(String) ⇒ Array

@jelaniwoods Yes, I like that!

jelaniwoods commented:
@raghubetina How do you think the method signature should look if the method can return a limited number of different types?

For example, `ActiveRecord_Relation#sum(Symbol)` can return a `String`,` Integer`, or `Float` depending on the column type. It can't return an `Array`, `Hash`, or `Date` so I didn't think using `Object` made the most sense.

`ActiveRecord_Relation#sum(Symbol) ⇒ Integer` or `⇒ Float` or `⇒ String`
or maybe
`ActiveRecord_Relation#sum(Symbol) ⇒ (Integer/Float/String)`
or even
`ActiveRecord_Relation#sum(Symbol) ⇒ Integer`
`ActiveRecord_Relation#sum(Symbol) ⇒ Float`
`ActiveRecord_Relation#sum(Symbol) ⇒ String`

jelaniwoods commented:
Additionally, how should method signatures for loops look? If the current method signature looks like
```bash
Integer#times do
# ruby code here
end
```

Do either of these look crazy?

```bash
Integer#times do
# ruby code here
end
⇒ Integer
```

```bash
Integer#times do
# ruby code here
end ⇒ Integer
```

raghubetina commented:
> How do you think the method signature should look if the method can return a limited number of different types?

Hmm I like the first two, 

`ActiveRecord_Relation#sum(Symbol) ⇒ Integer` or `⇒ Float` or `⇒ String`

and

`ActiveRecord_Relation#sum(Symbol) ⇒ (Integer/Float/String)`

or perhaps without parentheses,

`ActiveRecord_Relation#sum(Symbol) ⇒ Integer/Float/String`

Do you have any preference?

> how should method signatures for loops look?

I like this!

```
Integer#times do
  # ruby code here
end ⇒ Integer
```

or if we want to be more concise and look like the other ones, and drive home the idea that blocks are like arguments of code, maybe

```
Integer#times(Block) ⇒ Integer
```

`(Proc)` would be technically more accurate, but whatever. Also what about trying to sneak block variables into the signature?

```
Integer#times(Block(Integer)) ⇒ Integer
```

jelaniwoods commented:
@raghubetina I think I prefer `ActiveRecord_Relation#sum(Symbol) ⇒ Integer` or `⇒ Float` or `⇒ String`.

I
> or if we want to be more concise and look like the other ones, and drive home the idea that blocks are like arguments of code, maybe
> 
> ```
> Integer#times(Block) ⇒ Integer
> ```

I do like how concise this is. My concern is that it might not be clear what exactly a `Block` is. from the name. I feel like usually we say "don't forget to add the `end`!" and don't really name what we're really passing to the method. But maybe if we also add `Block` to the reference and link to it that would be better.

raghubetina commented:
or maybe now if we're trying to also show the block variable the more verbose one makes more sense:

```
Integer#times do |Integer|
  # ruby code here
end ⇒ Integer
```

```
Array#each do |Object|
  # ruby code here
end ⇒ Array
```

raghubetina commented:
> My concern is that it might not be clear what exactly a Block is.

I agree and it makes it seem like `Block` is actually a class, which it isn't.

When defining methods, we can name blocks with `def foo(&bar)` (and then `bar.call` rather than `yield`). Perhaps that's one way? `Integer#times(&block) ⇒ Integer`

But given [the idea of also including block variables in the signature](https://github.com/firstdraft/appdev-chapters/pull/4#issuecomment-523974129), I'm leaning towards the verbose version now.

raghubetina commented:
@jelaniwoods What do you think about putting the return value on the next line, to resemble `rails c`? Let's see how a few examples would look:

```
Array#each_with_index do |Object, Integer|
  # ruby code here
end

⇒ Array
```

```
ActiveRecord_Relation#sum(Symbol)
⇒ Integer/Float/String
```

```
String#split
String#split(String)
⇒ Array
```

jelaniwoods commented:
@raghubetina I like how having the return on a separate line looks for the loop example. I do miss how methods without blocks were on a single line and could be easily read quickly. 

When the method is often used to define a variable it feels less helpful to see the return value on a separate line. 

If I say, 
```ruby
a = "New string".split
```

I think it is easier to break down what `a` is with the one-line syntax.
```ruby
# String#split ⇒ Array
a = "New string".split ⇒ Array
# a is an Array
```

I don't think this makes much sense for loops though since you never really set a variable equal to the return value of a loop. What do you think?

raghubetina commented:
@jelaniwoods Okay, I buy that; we can go with the single-line style for methods that don't take blocks / where the return values are actually important.
